### DISCUSSION TITLE
Save and Load Methods

### DISCUSSION THREAD

**Discussion by mertbakir** [Nixtla/mlforecast#369] [Q&A] [answered]
Hello,

When we use lag features, I think we'll need the previous data at predict time. So, I save the model using save method. Then load it. Called the predict method passed forecast horizon and it just worked. How? Does save also saves the fitted data? Is this the right approach?

Is there a way to get prediction intervals after loading a saved model? Because it requires re-fitting the data.

Lastly, I would be happy if someone can simply explain what does the update method do and in which case we would use it. I didn't understand it from the docs.

**Comment by jmoralez** [comment, marked as answer]
>  Does save also saves the fitted data?

Yes, it saves the history

> Is there a way to get prediction intervals after loading a saved model?

No, since the scores aren't saved

> what does the update method do and in which case we would use it

This is kind of related to the first question. The model needs the history to compute the lag features, so if you trained a model a month ago and want to predict the next week you need to tell it what the values of your target were during this month. One option would be retraining with the original data plus the previous month's data, but you can also just update the target values (without modifying the model) so that it's able to predict the next week by using the values from the previous month (which you already know by that point).

**Reply by mertbakir** [reply]
Thank you very much for the quick reply. 

Just to clarify for the second question, if I need the prediction intervals as well as the point predictions, should I do batch prediction? save the predictions and the intervals into some data store, or is there another way I could follow for getting the predictions and intervals on request?

**Reply by jmoralez** [reply]
If you're predicting the same series (no transfer learning) then you could save the intervals artifacts and assign them back after loading, e.g.

```python
fcst = MLForecast(...).fit(...)
fcst.save('artifacts')
interval_attrs = {attr: getattr(fcst, attr) for attr in ('_cs_df', 'prediction_intervals')}
with open('artifacts/intervals.pkl', 'wb') as f:
    pickle.dump(interval_attrs, f)

# on a separate session
fcst2 = MLForecast.load('artifacts')
with open('artifacts/intervals.pkl', 'rb') as f:
    interval_attrs = pickle.load(f)
fcst2.__dict__.update(interval_attrs)
```

**Reply by mertbakir** [reply]
Gotcha, ty.
